File: app/src/main/java/com/popalay/yocard/ui/cards/CardsFragment.java
package com.popalay.yocard.ui.cards;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.popalay.yocard.R;
import com.popalay.yocard.data.models.Card;
import com.popalay.yocard.databinding.FragmentCardsBinding;
import com.popalay.yocard.ui.addcard.AddCardActivity;
import com.popalay.yocard.ui.base.BaseFragment;
import com.popalay.yocard.ui.base.ItemClickListener;
import com.popalay.yocard.utils.recycler.SimpleItemTouchHelperCallback;

import java.util.List;

import io.card.payment.CardIOActivity;
import io.card.payment.CreditCard;

public class CardsFragment extends BaseFragment implements CardsView,
        SimpleItemTouchHelperCallback.SwipeCallback, ItemClickListener<Card> {

    private static final int SCAN_REQUEST_CODE = 121;

    @InjectPresenter CardsPresenter presenter;

    private FragmentCardsBinding b;
    private CardsViewModel viewModel;

    public static CardsFragment newInstance() {
        return new CardsFragment();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        b = DataBindingUtil.inflate(inflater, R.layout.fragment_cards, container, false);
        return b.getRoot();
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initUI();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == SCAN_REQUEST_CODE) {
            if (data != null && data.hasExtra(CardIOActivity.EXTRA_SCAN_RESULT)) {
                CreditCard scanResult = data.getParcelableExtra(CardIOActivity.EXTRA_SCAN_RESULT);
                presenter.onCardScanned(scanResult);
            }
        }
    }

    @Override
    public void startCardScanning() {
        final Intent scanIntent = new Intent(getActivity(), CardIOActivity.class);
        startActivityForResult(scanIntent, SCAN_REQUEST_CODE);
    }

    @Override
    public void addCardDetails(CreditCard card) {
        startActivity(AddCardActivity.getIntent(getActivity(), card));
    }

    @Override
    public void showRemoveUndoAction(Card card) {
        Snackbar.make(b.listCards, R.string.card_removed, Snackbar.LENGTH_SHORT)
                .setAction(R.string.action_undo, view -> presenter.onRemoveUndo(card))
                .show();
    }

    @Override
    public void setItems(List<Card> items) {
        viewModel.setCards(items);
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder) {
        final Card card = viewModel.get(viewHolder.getAdapterPosition());
        presenter.onItemSwiped(card);
    }

    @Override
    public void onItemClick(Card item) {
        presenter.onCardClick(item);
    }

    private void initUI() {
        viewModel = new CardsViewModel();
        b.setModel(viewModel);
        b.setListener(this);

        new ItemTouchHelper(new SimpleItemTouchHelperCallback(this)).attachToRecyclerView(b.listCards);
        b.buttonAdd.setOnClickListener(v -> presenter.onAddClick());
    }
}

File: presentation/src/main/kotlin/com/popalay/cardme/screens/cards/CardsFragment.kt
package com.popalay.cardme.screens.cards

import android.arch.lifecycle.ViewModelProvider
import android.content.Intent
import android.os.Bundle
import android.support.v4.view.ViewCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.popalay.cardme.DataTransformers
import com.popalay.cardme.R
import com.popalay.cardme.base.BaseFragment
import com.popalay.cardme.databinding.FragmentCardsBinding
import com.popalay.cardme.screens.carddetails.CardDetailsActivity
import com.popalay.cardme.utils.DialogFactory
import com.popalay.cardme.utils.extensions.getDataBinding
import com.popalay.cardme.utils.extensions.getViewModel
import com.popalay.cardme.utils.extensions.hideAnimated
import com.popalay.cardme.utils.extensions.makeSceneTransitionAnimation
import com.popalay.cardme.utils.extensions.showAnimated
import com.popalay.cardme.utils.extensions.unsafeActivity
import com.popalay.cardme.utils.extensions.unsafeContext
import com.popalay.cardme.utils.recycler.SpacingItemDecoration
import io.card.payment.CardIOActivity
import io.card.payment.CreditCard
import io.reactivex.rxkotlin.addTo
import io.reactivex.rxkotlin.subscribeBy
import javax.inject.Inject

class CardsFragment : BaseFragment() {

	companion object {

		const val SCAN_REQUEST_CODE = 121

		fun newInstance() = CardsFragment()
	}

	@Inject lateinit var factory: ViewModelProvider.Factory
	@Inject lateinit var viewModelFacade: CardsViewModelFacade

	private lateinit var b: FragmentCardsBinding

	override fun onCreateView(
		inflater: LayoutInflater,
		container: ViewGroup?,
		savedInstanceState: Bundle?
	): View? {
		b = getDataBinding(inflater, R.layout.fragment_cards, container)
		b.vm = getViewModel(factory)
		initUI()
		return b.root
	}

	override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
		super.onActivityResult(requestCode, resultCode, data)
		if (requestCode == SCAN_REQUEST_CODE) {
			if (data != null && data.hasExtra(CardIOActivity.EXTRA_SCAN_RESULT)) {
				val scanResult = data.getParcelableExtra<CreditCard>(CardIOActivity.EXTRA_SCAN_RESULT)
				viewModelFacade.onCardScanned(DataTransformers.transform(scanResult))
			}
		}
	}

	override fun onResume() {
		super.onResume()
		b.buttonAdd.showAnimated()
	}

	override fun onPause() {
		super.onPause()
		b.buttonAdd.hideAnimated()
	}

	fun createCardDetailsTransition(activityIntent: Intent): Bundle {
		val position = viewModelFacade.getPositionOfCard(
			activityIntent.getStringExtra(CardDetailsActivity.KEY_CARD_NUMBER)
		)

		b.listCards.findViewHolderForAdapterPosition(position).itemView.run {
			val imageBackground = findViewById<View>(R.id.card_view)
			val imageType = findViewById<View>(R.id.image_card_type)
			val textTitle = findViewById<View>(R.id.text_title)
			val textNumber = findViewById<View>(R.id.text_number)
			val textHolderName = findViewById<View>(R.id.text_holder)

			return unsafeActivity.makeSceneTransitionAnimation(
				imageBackground to ViewCompat.getTransitionName(imageBackground),
				imageType to ViewCompat.getTransitionName(imageType),
				textTitle to ViewCompat.getTransitionName(textTitle),
				textNumber to ViewCompat.getTransitionName(textNumber),
				textHolderName to ViewCompat.getTransitionName(textHolderName)
			)
		}
	}

	private fun initUI() {
		b.listCards.addItemDecoration(SpacingItemDecoration.create {
			dividerSize = resources.getDimension(R.dimen.normal).toInt()
			showBetween = true
			showOnSides = true
		})

		viewModelFacade.doOnShowCardExistsDialog()
			.subscribeBy {
				DialogFactory.createCustomButtonsDialog(
					unsafeContext,
					R.string.error_card_exist,
					R.string.action_yes,
					R.string.action_cancel,
					onPositive = viewModelFacade::onWantToOverwrite,
					onDismiss = viewModelFacade::onShowCardExistsDialogDismiss
				)
					.apply { setCancelable(false) }
					.show()
			}.addTo(disposables)
	}
}
